### PR TITLE
access control engine and endpoint configuration

### PR DESCRIPTION
**Pull request by creechy**
**PR Checklist**

- [ ] A description of the changes is added to the description of this PR.
- [ ] If there is a related issue, make sure it is linked to this PR.
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added or modified a feature, documentation in `docs` is updated

**Description of changes**

### CODE REVIEW

File: server/src/main/java/io/unitycatalog/server/auth/annotation/AuthorizeExpression.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ *   <li>#authorizeAll(principal, resource, privilege ...)
+ * </ul>
+ *
+ * <p>And the context will include the following variables
+ *
+ * <ul>
+ *   <li>#principal - the (UUID) of the principal making the request
+ *   <li>#deny - constant evaluating to false
+ *   <li>#defer - constant evaluating to true - a marker to note authorization happens elsewhere
+ *   <li>#allow - constant evaluating to tru - always allow access
```
Comment (xx-db) [inline server/src/main/java/io/unitycatalog/server/auth/annotation/AuthorizeExpression.java]: nit: type `true`

File: server/src/main/java/io/unitycatalog/server/auth/decorator/UnityAccessDecorator.java [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+          } catch (IOException e) {
+            // IGNORE
+          }
+
+          // Unfortunately we don't appear to get a signal that we've got all the data, so we have to
+          // resort to attempting to parse the data whenever it _looks_ complete.
+          // TODO: try to optimize this using Jackson streaming or something else.
+          if (data.array()[data.array().length - 1] == '}') {
+            try {
+              Map<String, Object> payload = new ObjectMapper().readValue(dataStream.toByteArray(), new TypeReference<>() {
```
Comment (xx-db) [inline server/src/main/java/io/unitycatalog/server/auth/decorator/UnityAccessDecorator.java]: nit: will a static objectMapper work as well?

File: server/src/main/java/io/unitycatalog/server/auth/decorator/UnityAccessDecorator.java [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+      resourceIds.put(CATALOG, UUID.fromString(catalog.getId()));
+    }
+
+    if (resourceKeys.containsKey(CATALOG) && resourceKeys.containsKey(SCHEMA)) {
+      String fullName = resourceKeys.get(CATALOG) + "." + resourceKeys.get(SCHEMA);
+      SchemaInfo schema = SchemaRepository.getInstance().getSchema(fullName);
+      resourceIds.put(SCHEMA, UUID.fromString(schema.getSchemaId()));
+    }
+
+    // if only TABLE is specified, assuming its value is a full schema name (including catalog)
```
Comment (xx-db) [inline server/src/main/java/io/unitycatalog/server/auth/decorator/UnityAccessDecorator.java]: super nit: typo

**Review by xx-db — APPROVED** [review]
lgtm besides a few nits. Do we ever need to worry about having different keys for the same resourceType? like something that got passed in both in pathparam and payload.

File: server/src/main/java/io/unitycatalog/server/service/PermissionService.java [hunk 4]
```diff
@@ -26,37 +31,40 @@
...
+  public PermissionService(UnityCatalogAuthorizer authorizer) {
+    this.authorizer = authorizer;
+  }
 
   @Get("/{resource_type}/{name}")
   public HttpResponse getAuthorization(
       @Param("resource_type") ResourceType resourceType,
       @Param("name") String name,
-      @Param("principle") Optional<String> principle) {
+      @Param("principal") Optional<String> principal) {
```
Comment (xx-db) [inline server/src/main/java/io/unitycatalog/server/service/PermissionService.java:51]: thanks for fixing! 🫢

File: server/src/main/java/io/unitycatalog/server/auth/decorator/UnityAccessDecorator.java (hunk 2 above)
Comment (creechy) [reply server/src/main/java/io/unitycatalog/server/auth/decorator/UnityAccessDecorator.java]: good catch. i'll convert.

**Comment by creechy** [comment]
@xx-db 

> lgtm besides a few nits. Do we ever need to worry about having different keys for the same resourceType? like something that got passed in both in pathparam and payload.

I don't entirely envision that, but if it does happen, hopefully they would match and would just pick one.

**Pull request merged**